frontend/catalyst/ag_primitives.py

```diff
@@ -0,0 +1,140 @@
+    set_state: Callable[[Tuple], None],
+    symbol_names: Tuple[str],
+    _num_results: int,
+):
+    """An implementation of the Autograph 'if' statement. The interface is defined by Autograph,
+    here we merely provide an implementation of it in terms of Catalyst primitives."""
+
+    # Cache the initial state of all modified variables. Required because we trace all branches,
+    # and want to restore the initial state before entering each branch.
+    init_state = get_state()
+
+    @catalyst.cond(pred)
+    def functional_cond():
+        set_state(init_state)
+        true_fn()
+        results = get_state()
+        return assert_results(results, symbol_names)
+
+    @functional_cond.otherwise
+    def functional_cond():
+        set_state(init_state)
+        false_fn()
+        results = get_state()
+        return assert_results(results, symbol_names)
+
+    # Sometimes we unpack the results of nested tracing scopes so that the user doesn't have to
+    # manipulate tuples when they don't expect it. Ensure set_state receives a tuple regardless.
+    results = functional_cond()
+    if not isinstance(results, tuple):
+        results = (results,)
+    set_state(results)
+
+
+# Prevent autograph from converting PennyLane and Catalyst library code, this can lead to many
+# issues such as always tracing through code that should only be executed conditionally. We might
+# have to be even more restrictive in the future to prevent issues if necessary.
+module_allowlist = (
+    config.DoNotConvert("pennylane"),
+    config.DoNotConvert("catalyst"),
+) + config.CONVERSION_RULES
+
+
+def converted_call(fn, *args, **kwargs):
+    """We want Autograph to use our own instance of the AST transformer when recursively
+    transforming functions, but otherwise duplicate the same behaviour."""
+
+    # pylint: disable=protected-access
+    with Patcher(
+        (tf_autograph_api, "_TRANSPILER", catalyst.autograph._TRANSFORMER),
+        (config, "CONVERSION_RULES", module_allowlist),
+    ):
+        # We need to unpack nested QNode and QJIT calls as autograph will have trouble handling
+        # them. Ideally, we only want the wrapped function to be transformed by autograph, rather
+        # than the QNode or QJIT call method.
+
+        # For nested QJIT calls, the class already forwards to the wrapped function, bypassing any
+        # class functionality. We just do the same here:
+        if isinstance(fn, catalyst.QJIT):
+            fn = fn.user_function
+
+        # For QNode calls, we employ a wrapper to correctly forward the quantum function call to
+        # autograph, while still invoking the QNode call method in the surrounding tracing context.
+        if isinstance(fn, qml.QNode):
+
+            @functools.wraps(fn.func)
+            def qnode_call_wrapper():
+                return tf_converted_call(fn.func, *args, **kwargs)
+
+            new_qnode = qml.QNode(qnode_call_wrapper, device=fn.device, diff_method=fn.diff_method)
```

Comment:
Yeah this is a bit tricky, I included the ones I know we use (we ignore the rest I believe?), but from a maintainability point of view this is not great.
I didn't see an obvious way to copy all relevant properties, except maybe if we clone the object. Do you think `new_qnode = copy.copy(fn); new_qnode.func = qnode_call_wrapper` would be better? I'm not sure if there could still be some internal shared state though, `copy.deepcopy` might then be necessary 🤔

Reply:
Would a deepcopy always be desired? That would always copy the device.

I'm not totally sure what the desired behaviour is with `copy` vs `deepcopy`, but I do suspect that this operation is invoked infrequently enough that whatever overhead there will be from copying will probably not be a bottleneck.

Reply:
> Would a deepcopy always be desired? That would always copy the device.

I don't think it is necessarily desired, and probably overkill for most attributes. As you mention the device for example does not need copying, and this could impact performance (e.g. the lightning does (or did in the past) allocate large amounts memory upon initialization.

But a simple copy might lead to undesirable shared state between the two QNode objects.

Reply:
I see - with those things considered, the maintenance overhead of copying of specific attributes may be worth avoiding the risk of incorrect semantics from unexpected shared state or potential large overhead of extraneous device initialization.

At any rate, I don't see this as a blocker for this PR.

Reply:
Yes, I also have noticed this place. Looks not safe indeed. What about checking that the other arguments are the same? Or at least `assert str(new_qnode) == str(old_qnode)`?

Reply:
Are there specific attributes you are concerned about? Note that we don't use a majority of the QNode functionality.

The important thing I would say is that we don't affect the original QNode in anyway (as the user might still be using for other things), and we propagate the information the QNode provides to Catalyst. As far as I know there are only 3:
- the fact that wrapped function is a quantum function (we want to instantiate the corresponding primitive in the jaxpr)
- the differentiation method
- device information

But it's true that it's also easy to overlook additional properties in the future.

Reply:
I am talking from the purely formal point of view. My initial concern was that QNode is available for users and they might derive from it to use any number of custom attributes for their own purposes. Do we have an option to just replace the function inplace with a new one? It is true that I don't like the current solution but I am not insisting on any particular change.

Reply:
Hmm I see your concern. I'm fairly certain though that this QNode is not accessible to the user as we don't replace the original object they provided. The new QNode is merely used to invoke it, and should be discarded as we exit this scope (please correct me if I'm wrong).